### ISSUE TITLE
Sync identity provider groups and user deactivations from Active Directory using SCIM

### ISSUE THREAD

**Issue by rosszurowski** [tailscale/tailscale#979]
Several users have written in asking for the ability to sync user groups from their identity provider. This is really useful for providers like Okta, which are designed to provide this configuration to apps. We should save our users from this manual work and import it for them, probably using SCIM.

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_2j3pt)

**Comment by jack1902** [comment]
@rosszurowski wondering if this would be the start of breaking the ACL down into useable pieces?

Currently i see the syncing of groups and the ACL itself as two seperate pieces (even though currently, it is all defined within one ACL file)

Wondering if it would be possible to have the split so that, Groups and their members can be synced without having to change the ACL itself? For example, i might want to control what groups have access to other groups but not care about who is in those groups as that is handled by Okta. Currently, i'd have to ensure that i have the latest ACL from the API before i make any updates to the ACL itself so that the `Groups` section isn't completely wiped out by one update call

**Comment by rosszurowski** [comment]
@jack1902 good point. I'm not sure the exact form this would take. @apenwarr has thought a lot about our ACL file and may have opinions here.

In general, I think we do want to keep the single ACL file as a "source of truth" for what's allowed on a domain so it can be updated via API, version-controlled, etc. But even if a single file is the source of truth, that doesn't mean we can't provide more convenient interfaces on top of it. We may, for example, introduce separate API endpoints to allow editing just `"ACLs"` without touching the `"Groups"` or `"Hosts"`. Similarly, we'll probably introduce a UI for managing user groups that shouldn't affect other sections.

**Comment by apenwarr** [comment]
When we get the auto-syncing working, the plan is to use Okta SCIM to sync into something like an autogroup: section that can be referred to from the ACL file, but not actually be listed in the ACL file. Then the existing statically-defined groups can be used a lot less often, and thus the ACL file itself will change a lot less often.

Separately, as we're working on an API for modifying ACLs, we should have it accept an atomic "overwrite-if-equals" kind of operation, which will reject changes to the ACLs unless the change request also includes eg. the ETag of the ACLs that are supposed to be there before the change. If the ETag doesn't match, then someone else got there before you, sorry, try again. Like an opportunistic transaction in a database. That should safely cover all cases, although it would be a little annoying for a client to implement sometimes.

**Comment by DentonGentry** [comment]
Summarizing some internal discussions about creation of autogroups using SCIM:
- Until something automatic is working, the API can also be used to automate rule creation: http://github.com/tailscale/tailscale/blob/main/api.md
- Okta and AzureAD are the two highest priority

**Comment by DentonGentry** [comment]
For those following along, work on SCIM for Okta and Azure is underway.

**Comment by DentonGentry** [comment]
SCIM design review happened this week.

**Comment by hrzbrg** [comment]
Are there updates on this topic? SCIM and groups would significantly reduce the manual administrative overhead for us.

**Comment by DentonGentry** [comment]
Group and user suspension synchronization for Okta exited Alpha yesterday: https://tailscale.com/blog/sync-okta-groups/

It is now available in Beta for anyone to try. Please open new issues to report problems you encounter, don't add it here. This issue tracks the overall effort.

**Comment by hopinto** [comment]
Would OneLogin SCIM support make sense from a business perspective?

**Comment by DentonGentry** [comment]
I would expect to eventually add OneLogin support as well if it is possible to do so, but expect Azure to come next before Onelogin work would start.

**Comment by owenda7** [comment]
Are there any time estimates for AzureAD SCIM Beta release?

**Comment by DentonGentry** [comment]
I promise that if progress is made on this, we'll update the bug. Much of the work done on SCIM for Okta will be directly applicable.

**Comment by nerfpops** [comment]
Shouldn't an open standard like SCIM just work with various providers? How come a server must have knowledge about which provider is provisioning users (besides a shared key of some sort)?

**Comment by DentonGentry** [comment]
Having an implementation working with one provider helps a great deal, but rarely works immediately until the standard has become extremely widely implemented (which SCIM has not). There is debugging, and adapting to the interpretations each provider makes of what the standard says to do.

**Comment by josedev-union** [comment]
Could you share the status of  AzureAD SCIM release?

**Comment by DentonGentry** [comment]
When we have progress to report it will be posted here.
The internal work being done is not to the point of posting here yet.

**Comment by DentonGentry** [comment]
If someone wishes to construct a bespoke automation to do this, the HuJSON Patch function may be useful. An example of the HuJSON patching can be found at https://github.com/DentonGentry/tailscale-aws-host-acl-updater/blob/main/main.go

1. Fetch the current complete ACL JSON, including the ETag header: https://github.com/DentonGentry/tailscale-aws-host-acl-updater/blob/739713db985399ac1939b7934d4aea45fab56c12/acls.go#L17
1. Use HuJSON to patch in an update like updating the IP address in a Hosts entry, without having to parse the entire JSON: https://github.com/DentonGentry/tailscale-aws-host-acl-updater/blob/739713db985399ac1939b7934d4aea45fab56c12/acls.go#L86
1. Only write the ACL JSON back if the ETag has not changed in the meantime: https://github.com/DentonGentry/tailscale-aws-host-acl-updater/blob/739713db985399ac1939b7934d4aea45fab56c12/acls.go#L54
1. If the If-Match fails because the ETag did change, loop and try again: https://github.com/DentonGentry/tailscale-aws-host-acl-updater/blob/739713db985399ac1939b7934d4aea45fab56c12/acls.go#L72
    It should probably add a random sleep in the retry case, to avoid livelock between two clients, but currently does not.

**Comment by mayakacz** [comment]
Development on user & group provisioning (SCIM) for Azure AD is underway. When we have further progress to report, we will post it here.

**Comment by pcibraro** [comment]
We have an existing subscription In Tailscale, and we would like to try to SCiM APIs directly without using Azure AD or Okta. We have a customer adapter we wrote for our identity provider. Are those APIs available for any customer to use?

**Comment by DentonGentry** [comment]
There is an API to update the ACL file where groups are defined: https://github.com/tailscale/tailscale/blob/main/api.md#update-policy-file

The flow would be:
1. Fetch the current Tailscale ACLs.
1. Apply the updated group definitions. HuJSON has a Patch operation which can do this without needing to fully parse the original ACLs.
1. POST back to the API to write the new ACLs using If-Match to determine if there were any intervening writes.

There is an example of doing this in https://github.com/DentonGentry/tailscale-aws-host-acl-updater, in acls.go.

**Comment by pcibraro** [comment]
Yes, I am familiar with the ACL Api. We were planning to use that one until I saw you had a beta version for SCIM. The thing with the ACL Api is that we require writing custom code just to integrate with Tailscale when we had a SCIM implementation already in place for pushing data.

**Comment by DentonGentry** [comment]
Every identity provider we've worked with has required substantial effort from us. SCIM is a large spec, of which every provider implements a different subset. The Patch operation has required new work as each IdP uses it quite differently, their subset of attributes is different each time, and their interpretation of what Suspended and other states mean has varied greatly. As such, we don't make the SCIM API publicly available. It has required us to do substantial custom work each time.

The mechanism we have available is https://github.com/tailscale/tailscale/blob/main/api.md#update-policy-file

**Comment by pcibraro** [comment]
Sounds good! I understand the issue. Thanks.

**Comment by plett** [comment]
Good news! Syncing Azure AD users and groups to Tailscale went into beta last week: https://tailscale.com/kb/1249/sso-azure-ad-scim/

I have been subscribed to this issue waiting for the promised updates on progress from Tailscale staff but then spotted this KB article in their changelog feed, so thought I would let everyone else who has been watching this issue know about it.

**Comment by TheMrDrProf** [comment]
Curious if anyone has got this setup and working. Super happy to see it implemented, however groups are not syncing. We configured precisely following the Tailscale docs, other than we did not create the Tailscale Enterprise App since it already existed as our auth method to Tailscale; otherwise configured precisely following docs.

SCIM in Azure AD shows provisioning cycles running. I only have 5 groups directly assigned to the Tailscale app for sync. There are 3 users that are members of some number of these groups. SCIM provisioned the 3 users (which already existed in Tailscale), but is not syncing any groups. Reviewing the provisioning logs, all groups that are queued for sync are skipped:
![Web capture_8-9-2023_193559_portal azure com](https://github.com/tailscale/tailscale/assets/25834380/d4ff510c-ea04-484c-bdd4-6e65aa85ea9c)

On Tailscale side, it shows successful syncs, but no groups:
![Web capture_8-9-2023_193644_login tailscale com](https://github.com/tailscale/tailscale/assets/25834380/9e662276-ffaa-4211-b6df-9c3010b9d32e)

Anyone able to provide guidance on this?

**Comment by hazcod** [comment]
@TheMrDrProf we have the exact same situation and issue, no resolution from support yet.

**Comment by crawshaw** [comment]
Hello. We have identified an issue with Azure group sync and are working with Microsoft to get it resolved. I don’t have an ETA yet, but we will update this thread when we know more.

**Comment by deandre** [comment]
To expand on @crawshaw's comment, there was a bug in how Azure AD enabled group sync for Tailscale, which uncovered a larger group sync issue for Azure AD. Azure AD is working on this issue, and specifically republishing the Tailscale app. We expect to confirm a fix this week, and republish the app in the coming ~2 weeks.

When the app is republished, group sync should be enabled by default and start working for any existing installations of the app.

Meanwhile, if necessary, you can install Tailscale as a custom app for group sync. This is not recommended as it will need to be replaced once the updated app is published.

**Comment by TheMrDrProf** [comment]
This appears to now be working for me. Groups have appeared in the SCIM settings in Azure AD:
![image](https://github.com/tailscale/tailscale/assets/25834380/74db92a7-dfca-45c1-bee0-e0925466f57a)

And I now have groups available in Tailscale: (Ignore the [WARP], we came from Cloudflare)
![image](https://github.com/tailscale/tailscale/assets/25834380/e264d044-179c-4756-b994-cf277a368a5b)
![image](https://github.com/tailscale/tailscale/assets/25834380/d69581b4-0ed2-4656-8228-39f0a78392b5)

**Comment by ketumbra** [comment]
This works now right? Can this be closed?

**Comment by andrew-d** [comment]
This is now implemented and working; if anyone has issues with Azure SCIM sync, please reach out to [Tailscale support](https://tailscale.com/contact/support#support-form) for more help!
